zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/CommitProcessor.java

```diff
@@ -621,6 +622,20 @@ public void shutdown() {
             workerPool.join(workerShutdownTimeoutMS);
         }
 
+        try {
+            this.join(workerShutdownTimeoutMS);
+        } catch (InterruptedException e) {
+            LOG.warn("Interrupted while waiting for CommitProcessor to finish");
+            Thread.currentThread().interrupt();
+        }
+
+        if (this.isAlive()) {
+            LOG.warn("CommitProcessor does not shutdown gracefully after "
+                    + "waiting for {} ms, exit to avoid potential "
+                    + "inconsistency issue", workerShutdownTimeoutMS);
+            System.exit(ExitCode.SHUTDOWN_UNGRACEFULLY.getValue());
```

Comment:
Why aren't we using 'halt' ?

'exit' triggers shutdown hooks and they are sometimes source of deadlocks (I don't know in ZK code)

Reply:
I believe we don't use shutdown hook in server code.

Reply:
@eolivelli we still try to halt before checking and exit here, but in case the CommitProcessor thread itself cannot shutdown within given time, we need to quit safely. And as @hanm mentioned we don't use shutdown hook here.